crates/sui-core/src/authority_server.rs

```diff
@@ -672,6 +672,37 @@ impl ValidatorService {
             })
     }
 
+    fn transaction_validity_check(
+        epoch_store: &Arc<AuthorityPerEpochStore>,
+        transaction: &SenderSignedData,
+    ) -> SuiResult<()> {
+        let config = epoch_store.protocol_config();
+        // CRITICAL: DO NOT ADD ANYTHING BEFORE THIS CHECK.
+        // This must be the first thing to check because the transaction may not even be valid to
+        // access for any other checks.
+        transaction.validity_check(config, epoch_store.epoch())?;
+
+        if !config.zklogin_auth() && transaction.has_zklogin_sig() {
+            return Err(SuiError::UnsupportedFeatureError {
+                error: "zklogin is not enabled on this network".to_string(),
+            });
+        }
+
+        if !config.supports_upgraded_multisig() && transaction.has_upgraded_multisig() {
+            return Err(SuiError::UnsupportedFeatureError {
+                error: "upgraded multisig format not enabled on this network".to_string(),
+            });
+        }
+
+        if !epoch_store.randomness_state_enabled() && transaction.uses_randomness() {
+            return Err(SuiError::UnsupportedFeatureError {
+                error: "randomness is not enabled on this network".to_string(),
+            });
+        }
```

Comment:
should these checks be moved to TransactionData::check_version_supported? not sure myself tbh

Reply:
We probably should. It does a bunch of feature checks there too. Would be good to put them in once place.